### PR TITLE
fix/resolve endless progress loading

### PR DESCRIPTION
**Pull request by Efhemo**
##Fix
Progressbar keeps loading when there is a network failure on fetching fetchPokemonList

### Types of changes
What types of changes does your code introduce?
- [ ] Bugfix

### CODE REVIEW

File: app/src/main/java/com/skydoves/pokedex/repository/MainRepository.kt [hunk 1]
```diff
@@ -75,5 +77,5 @@ class MainRepository @Inject constructor(
       emit(pokemonDao.getAllPokemonList(page))
       onSuccess()
     }
-  }.onStart { onStart() }.flowOn(Dispatchers.IO)
+  }.onStart { onStart() }.flowOn(Dispatchers.IO).onCompletion { onDone() }
```
Comment (skydoves) [inline app/src/main/java/com/skydoves/pokedex/repository/MainRepository.kt:80]: As I mentioned in the comment, let's change this point using the `onComplete()`.

```kotlin
}.onStart {
    onStart()
  }.onCompletion {
    onComplete()
  }.flowOn(Dispatchers.IO)
```

File: app/src/main/java/com/skydoves/pokedex/repository/MainRepository.kt [hunk 2]
```diff
@@ -44,7 +45,8 @@ class MainRepository @Inject constructor(
     page: Int,
     onStart: () -> Unit,
     onSuccess: () -> Unit,
-    onError: (String?) -> Unit
+    onError: (String?) -> Unit,
+    onDone: () -> Unit
```
Comment (skydoves) [inline app/src/main/java/com/skydoves/pokedex/repository/MainRepository.kt:49]: How do you think just change `onSuccess` to `onComplete` and remove the `onDone`?
I think actually the role of the `onSuccess` is only making the `isLoading` false.

File: app/src/main/java/com/skydoves/pokedex/ui/main/MainViewModel.kt [hunk 3]
```diff
@@ -55,8 +55,9 @@ class MainViewModel @Inject constructor(
       mainRepository.fetchPokemonList(
         page = page,
         onStart = { isLoading = true },
-        onSuccess = { isLoading = false },
-        onError = { toastMessage = it }
+        onSuccess = { },
+        onError = { toastMessage = it },
+        onDone = { isLoading = false },
```
Comment (skydoves) [inline app/src/main/java/com/skydoves/pokedex/ui/main/MainViewModel.kt:60]: And then, we can change this like the below.

```kotlin
onComplete = { isLoading = false },
onError = { toastMessage = it }
```

**Review by skydoves — CHANGES_REQUESTED** [review]
Thanks for your PR!
I think using the `onComplete` is a nice idea 💡  and I left some comments.

File: app/src/main/java/com/skydoves/pokedex/repository/MainRepository.kt (hunk 1 above)
Comment (Efhemo) [reply app/src/main/java/com/skydoves/pokedex/repository/MainRepository.kt:80]: yes. Nicely format

File: app/src/main/java/com/skydoves/pokedex/repository/MainRepository.kt (hunk 2 above)
Comment (Efhemo) [reply app/src/main/java/com/skydoves/pokedex/repository/MainRepository.kt:49]: yes. We can do that. the onSuccess is redundant.

**Review by skydoves — APPROVED** [review]
LGTM :)

**Pull request merged**